flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetSingleRowJoin.scala

```diff
@@ -46,6 +47,7 @@ class DataSetSingleRowJoin(
     rowRelDataType: RelDataType,
     joinCondition: RexNode,
     joinRowType: RelDataType,
+    joinType: JoinRelType,
```

Comment:
You are right, the check in `DataSetSingleRowJoinRule` should already prevent this case. However, I think it is good to have an alternative check here to be on the safe side. The translation process might be changed be adding or modifying rules. So ideally, this check will never trigger but if it catches something, this is a strong indication that the optimization produced a wrong results.